### PR TITLE
Cache .mat file reads

### PR DESCRIPTION
**Pull request by halebodc**
Reading .mat files from a network drive can be a slow process, and has potentially many repetitions over consecutive profiles. Since this use case is common, this PR attempts to address this issue by caching .mat file read (and data transforms). Currently we have limited the cache size to 50 as the .mat files may be ~50mb, however this could be optimised further.

### CODE REVIEW

**Comment by gmaze** [comment]
Hi @halebodc , just to let you know that work at euroargodev/argopy/issues/141 may have some link with this PR

**Comment by halebodc** [comment]
> Hi @halebodc , just to let you know that work at [euroargodev/argopy/issues/141](https://github.com/euroargodev/argopy/issues/141) may have some link with this PR

Hi @gmaze, thanks very much for pointing out the improvements you're making to `argopy`. It looks like it will be beneficial, particularly when we reach the point of allowing for input data in formats other than `.mat`

**Comment by gmaze** [comment]
thanks @halebodc 

> allowing for input data in formats other than `.mat`

Note that argopy will be able to do both:
- Return preprocessed data as xarray dataset
- Save preprocessed data in Matlab files to be used wherever needed

File: pyowc/data/fetchers.py [hunk 1]
```diff
@@ -19,6 +19,46 @@
...
+    return read_mat(filepath)
+
+
+@lru_cache(maxsize=50)
+def read_mat_bottle(filepath):
+    """Load and transform a bottle .mat file."""
+    data = read_mat(filepath)
+
+    # if data dimensions don't match, transpose to avoid indexing issues
+    if (data['lat'].size == data['pres'].size and data['lat'].shape[1] == data['pres'].shape[1]):
```
Comment (Fakas) [inline pyowc/data/fetchers.py:34]: Can these parentheses be dropped?

File: pyowc/data/fetchers.py [hunk 2]
```diff
@@ -19,6 +19,46 @@
...
+    return data
+
+
+@lru_cache(maxsize=50)
+def read_mat_argo(filepath, pa_float_name):
+    """Load and transform an Argo .mat file."""
+    data = read_mat(filepath)
+    # remove the argo float being analysed from the data
+    not_use = []
+    for i in range(0, data['lat'].__len__()):
```
Comment (Fakas) [inline pyowc/data/fetchers.py:49]: Why the use of `__len__()` here? Could we use `len(data['lat'])` or `data['lat'].size`?

**Review by Fakas — CHANGES_REQUESTED** [review]
Only a couple of minor comments.

File: pyowc/data/fetchers.py (hunk 2 above)
Comment (halebodc) [reply pyowc/data/fetchers.py:49]: The code itself was simply moved into this function for caching. I avoided making any other changes to minimise the chance of introducing other regressions. I agree that this usage is non-standard and we should remove it, I'm just not sure if we want to do it as part of this PR.

File: pyowc/data/fetchers.py (hunk 1 above)
Comment (halebodc) [reply pyowc/data/fetchers.py:34]: See https://github.com/euroargodev/argodmqc_owc/pull/72#discussion_r744596357

**Pull request merged**